File: day01/main.py
data = list()

with open("data.txt", "r") as f:
    temp = list()
    for line in f:
        line = line.replace("\n","")
        if line == "----":
            data.append(temp.copy())
            temp.clear()
            continue
        temp.append(line)
data = data[1:]

data = [
    [v[0],int(v[1])] 
    for v in data
]

numbers = [
        "one", "two", "three",
        "four", "five", "six",
        "seven", "eight", "nine"
        ]

count = 0

for string, val in data:
    first_digit = 0
    second_digit = 0

    i = 0
    found = False
    while i < len(string) and not found:
        if string[i].isdigit():
            first_digit = int(string[i])
            break

        for x in range(len(numbers)):
            if string[i:].startswith(numbers[x]):
                found = True
                first_digit = x+1
                break

        i += 1

    j = len(string) - 1
    found = False
    while j >= 0 and not found:
        if string[j].isdigit():
            second_digit = int(string[j])
            break

        for x in range(len(numbers)):
            if string[:j+1].endswith(numbers[x]):
                found = True
                second_digit = x+1
                break

        j -= 1

    calibration = first_digit*10 + second_digit
    if calibration != val:
        print(f"{string}: {calibration}, {val}")
    count += calibration

# print(count)

File: day01/src/main.rs
use std::fs::File;
use std::io::prelude::*;

fn task01() -> std::io::Result<()> {
    let contents = std::fs::read_to_string("input.txt")?;
    let sum = contents.lines().fold(0, |sum, x| {
        sum + 10 * x.chars().find_map(|c| c.to_digit(10)).unwrap()
            + x.chars().rev().find_map(|c| c.to_digit(10)).unwrap()
    });

    println!("{}", sum);

    Ok(())
}

fn task02() -> std::io::Result<()> {
    let numbers = vec![
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];

    let contents = std::fs::read_to_string("input.txt")?;


    let sum = contents.lines().fold(0, |sum, x| {
        let l_digit = x.find(|c: char| c.is_digit(10)).unwrap();
        let r_digit = x.rfind(|c: char| c.is_digit(10)).unwrap();

        let (left, _) = x.split_at(l_digit);
        let (_, right) = x.split_at(r_digit);

        // index, value
        let mut left_n = (
            l_digit,
            x.chars().nth(l_digit).unwrap().to_digit(10).unwrap(),
        );
        let mut right_n = (-1, x.chars().nth(r_digit).unwrap().to_digit(10).unwrap());

        for (i, number) in numbers.iter().enumerate() {
            let i = (i + 1) as u32;

            if let Some(pos) = left.find(number) {
                // potential improvement: #![feature(let_chains)]
                // https://github.com/rust-lang/rust/issues/53667
                if left_n.0 > pos {
                    left_n = (pos, i);
                }
            };

            if let Some(pos) = right.rfind(number) {
                let pos = pos as isize;
                if right_n.0 < pos {
                    right_n = (pos, i);
                }
            };
        }

        sum + 10 * left_n.1 + right_n.1
    });

    println!("{}", sum);
    Ok(())
}

fn main() -> std::io::Result<()> {
    task01()?;
    task02()?;

    Ok(())
}